### ISSUE TITLE
Add flag for ai-controlled players

### ISSUE THREAD

**Issue by oy** [teeworlds/teeworlds#1390]
Add a flag for ai-controlled players (client, server) so you can filter them within the server browser (similar to spectators).

**Comment by pielover88888** [comment]
I think this would be good, as of 0.6 I've noticed when I want to play CTF, and nobody is on, it looks as if there are 4 players on that one or two bot servers. Would be nice if 0.7 didn't have that small UX issue

**Comment by heinrich5991** [comment]
Stitch626 told me that we could mimick UT3's display for bots: <num bots>+<num players>/<max players (not including bots)>.

**Comment by heinrich5991** [comment]
![Screenshot](http://i259.photobucket.com/albums/hh305/HideInLight/ServerBrowserSuggestions.jpg)

**Comment by oy** [comment]
Yeah, seems like a good idea.

**Comment by Stitch626** [comment]
As little info, the value after + is the bot count ;)

**Comment by oy** [comment]
Yeah, so you also notice in while playing.
Like an antenna? :D

**Comment by android272** [comment]
There could be a bot_x skin like we do with ninjas.

**Comment by oy** [comment]
Yeah could be a total skin or just a part - we're flexible there with the new skin system :)

**Comment by Dune-jr** [comment]
> Yeah, so you also notice in while playing.
> Like an antenna? :D

Ouch, not sure if that's a good idea to edit the actual in-game skin.

Some bot gametypes like to provide specific skins (like zombies or what not). This would kinda break the functionality.

**Comment by LordSk** [comment]
How is this going to work, exactly? Is this an opt-in feature, like signaling players that X client is a bot?

**Comment by oy** [comment]
@LordSk 
That's supposed for server sided bots. Indication for players that these are bots:
- in the browser: be able to filter them, dunno if they should be displayed  individually
- ingame some indication in the scoreboard and/or for ingame skin?

**Comment by LordSk** [comment]
Oh yeah I understood that part. I meant if I ever made said ai-controlled bots, how would clients know to identify them?

**Comment by oy** [comment]
Would add a player flags https://github.com/teeworlds/teeworlds/blob/master/datasrc/network.py#L9 for that. 
Dunno if there should be an opt-in feature for client bots which connect to a server? guess you were refering to these?

**Comment by android272** [comment]
> Some bot gametypes like to provide specific skins (like zombies or what not). This would kinda break the functionality.

Well for game types that do this they could give their zombies a zombie_x skin but still have bots. But for other game types that just have normal bots they can use the bot_x for the bot skin.

Are skins downloadable? I think if a server has a map with assets you don't have you first download those assets before you can join the server right? is this done with skins as well?

**Comment by Dune-jr** [comment]
> [quakenet] \<Oy\> some bot flag might be good for the server browser so you can identify them, when they aren't filtered?
> [quakenet] \<Dune\> yeah
> [quakenet] \<Dune\> probably should replace the country flag
> [quakenet] \<Dune\> (the flag makes no sense for bots)
> [quakenet] \<Oy\> yeah

Anyone on it? :)

**Comment by android272** [comment]
Hay Oy, why are skins not downloadable? Would make a lot of sense for game types with special skins.

**Comment by Dune-jr** [comment]
I see one reason: penis

**Comment by Stitch626** [comment]
Nice reasons, Dune.

**Comment by android272** [comment]
> I see one reason: penis
> Second reason: invisible/hard to aim tees

But if you are downloading it from the server this may not be a problem. I don't know why a server maintainer would put an invisible skin on their server. I guess someone could add a penis skin.

also as it stands there is nothing stopping someone from making a map called ctf1 replacing the trees from grass_dodads with penis' and naming it grass_dodads2 so that everyone has to download it.

**Comment by Stitch626** [comment]
There is a mapchecker which checks for the mapcrc (of vanilla ones) if I remember allright.

**Comment by Dune-jr** [comment]
@android272 then you would like a skin downloader, not from other clients, but from the server? I mean, sure that could be done, but it sounds like a bunch of resources to download for a limited use, since you still can't see others' skins, doesn't it?

Regarding the map penis issue: well, yeah. Like @Stitch626 says, if you play on pure maps only, you are protected, though.

**Comment by android272** [comment]
Yes, that is what I would like. Download from the server. 

What do you mean you can't see others skins? So if I download a database of skins from some sight when I play with a player that uses a skin in that database I still won't see them using it? What is the point of having a custom skin if nobody can see you using it?

**Comment by Dune-jr** [comment]
@android272 Yeah, you would, skins get matched by name.

But that means, if you want custom skins from the server, that players would have to select custom skins on each server? Because there would be no reason for the server to have their skin.

**Comment by Skaymouse** [comment]
maybe we could add bot / zombie / alien parts to the vanilla client (?) also my idea to make a flag is a light gray flag with some futuristic logo in it, or a logo related to AI, like a roboic brain for example

**Comment by Zatline** [comment]
@oy Dune asked me to post skin parts here .. 
please notice this parts are 32 px heiger than regular parts.
![test_bot](https://user-images.githubusercontent.com/344608/54084712-79a92780-4334-11e9-8a0f-abb4e738443a.png)
^ this is to help align it to account for the 32 px height difference, 
red stroke should cover the tees inner outline
![x_bot](https://user-images.githubusercontent.com/344608/54084706-66965780-4334-11e9-84f3-770fb4e711a0.png)
^ this one is to be used once perfectly aligned

![ec8cc1fee8f71c6eeaae5fe8a47c059b](https://user-images.githubusercontent.com/344608/54084893-5af86000-4337-11e9-87d0-e1aa54a2bdc7.png)
^ I imagined bots to have a metalic look with randomized tattoos and eye/antenna color
colors used are #f9f9f9ff (white) and #b5b5b5ff (grey) for body/tattoos and vise versa

**file instructions:**
![08d6e407f44bb466c9b8ee4db77d9706](https://user-images.githubusercontent.com/344608/54084799-a3af1980-4335-11e9-9589-e4f50eefd620.png)

the graphic consists of 3 layer types
- a colorable layer (above the skin)
- a no color layer (above the skin)
- a backround layer (behind the skin)

the idea is to additionally lay the ground for overall cosmetics

**Comment by oy** [comment]
@Zatline looks really good :)
The colour parts might be too small for team based gametypes, to easily distiguish the team?

**Comment by Dune-jr** [comment]
This is what it looks like: https://gfycat.com/UnrulyPlayfulDogwoodtwigborer.gif

Not too good yet I guess? (didn't do the eyes)
![image](https://user-images.githubusercontent.com/355114/55103536-f116df00-50c8-11e9-95dc-fc8f457700d9.png)

**Comment by Dune-jr** [comment]
An attempt: 
![image](https://user-images.githubusercontent.com/355114/55173431-fed96c80-517b-11e9-9451-90002822f917.png)

The right thing looks a bit off to me.

**Comment by Zatline** [comment]
use the graphic with the red circle and align that to the tees body until red covers all of the blacks.

**Comment by Dune-jr** [comment]
@Zatline I don't think I can align much in the code. I am simply printing this image:
![image](https://user-images.githubusercontent.com/355114/55196707-7fb25b80-51b0-11e9-90c4-dd242e7c70ab.png)

on top of the tree. Maybe the gfx need to be adjusted to take in account for the outline?

**Comment by Dune-jr** [comment]
I wonder if the colors should be randomly assigned to each bot and stored in the client info, or depend on the bot ID.

![image](https://user-images.githubusercontent.com/355114/55559949-87fa2180-56ef-11e9-83a7-3a69a224274a.png)

**Comment by Sonix-** [comment]
> I am simply printing this image on top of the tree. 
What about printing the image underneath the body? If possible

**Comment by AssassinTee** [comment]
> I wonder if the colors should be randomly assigned to each bot and stored in the client info, or depend on the bot ID.
> 
> ![image](https://user-images.githubusercontent.com/355114/55559949-87fa2180-56ef-11e9-83a7-3a69a224274a.png)

From experience i can tell, that if you want distinguishable colors, you should simply iterate over the HUE. But I think this shouldn't be needed: Blue for Team Blue, red for Team Red and simply white or gray for DM should do the job.

I would still store it in the client info (7th body part?), because mod-programmer would request this feature anyways and you could better distinguish shop-keeper from quest-npc (or similar).

**Comment by Dune-jr** [comment]
Tried to make a gif:
![Webp net-gifmaker (1)](https://user-images.githubusercontent.com/355114/55623249-b553d780-57a2-11e9-81b6-bb8cc4c2ac23.gif)


Image on top:
![screenshot_2019-04-05_12-58-39](https://user-images.githubusercontent.com/355114/55623268-c7ce1100-57a2-11e9-8307-0184e873eac0.png)

Image underneath:
![screenshot_2019-04-05_12-58-56](https://user-images.githubusercontent.com/355114/55623270-c7ce1100-57a2-11e9-8c78-d9a2aacd23cf.png)


> From experience i can tell, that if you want distinguishable colors, you should simply iterate over the HUE. But I think this shouldn't be needed: Blue for Team Blue, red for Team Red and simply white or gray for DM should do the job.

To clarify, I'm going with Zatline's colors:

```cpp

				static const unsigned char aBotColors[][3] = {
					{0xff,0x00,0x00},
					{0xff,0x66,0x00},
					{0x4d,0x9f,0x45},
					{0xd5,0x9e,0x29},
					{0x9f,0xd3,0xa9},
					{0xbd,0xd8,0x5e},
					{0xc0,0x7f,0x94},
					{0xc3,0xa2,0x67},
					{0xf8,0xa8,0x3b},
					{0xcc,0xe2,0xbf},
					{0xe6,0xb4,0x98},
					{0x74,0xc7,0xa3},
				};
```

The question was only about whether it should be randomly generated or always be the same for a given client ID. I don't think it matters too much though.

**Comment by oy** [comment]
Randomly would probably be the best.

**Comment by Dune-jr** [comment]
> Randomly would probably be the best.

Alright, so this implies adding a `BotColor` field in the client struct.

**Comment by Sonix-** [comment]
Looks better to me, placed underneath. 👍

**Comment by Zatline** [comment]
what about the metallic body?

**Comment by Dune-jr** [comment]
@Zatline I'm afraid people would be unhappy about being unable to set bots to a particular skin or color, or have any sort of customizable attribute.

I don't know of a good solution, unfortunately. Should we force this metallic color palette? Modders may be reluctant to set it at all if we do that.

**Comment by oy** [comment]
The difference might be too small if all bots use that same metallic color - especially in teamplay.

**Comment by Dune-jr** [comment]
We still need a countryflag for bots in the server browser. Some external tools like status.tw are also waiting on it. Most straightfoward way would be to mix up
![default](https://user-images.githubusercontent.com/355114/64465493-3d049000-d10d-11e9-891a-0a8193ad3468.png)
and
![image](https://user-images.githubusercontent.com/355114/64465521-5c9bb880-d10d-11e9-89e6-0709fa6f41ef.png)
in one image.
I would do it but it'd probably look awful. https://status.tw uses this:
![image](https://user-images.githubusercontent.com/355114/64465570-af757000-d10d-11e9-9ba4-cea2fdf924d0.png)

**Comment by Dune-jr** [comment]
Another issue is that in 0.6 it was forbidden for server operators to display bots in the server info. In 0.7 the corresponding flag was added, and there is a filter to hide bots in the server list, but by default it is disabled and bots are not marked as bots in the player list (that part would be fixed by the countryflag). According to the current logic, I think it would also be possible to start a server with 64 bots on it and only disconnect if a real player wants to join. The whole thing might miss the logic behind this rule change.

If I remember correctly, I was the one who asked for bots to be shown by default. The logic behind this was that players should discover that there are servers with bots that they can play on, before they decide they are not interested in them and hide them. So in the end I'm not sure what the best course of action is here. Opinions?

**Comment by AssassinTee** [comment]
Do it as @heinrich5991 suggested, split the player count in player and bot count. This way the people see, that there are bots in this server and are not tricked into thinking that there are players

**Comment by nheir** [comment]
> If I remember correctly, I was the one who asked for bots to be shown by default. The logic behind this was that players should discover that there are servers with bots that they can play on, before they decide they are not interested in them and hide them. So in the end I'm not sure what the best course of action is here. Opinions?

In complement with a separate bot count, we can tweak the sort such that it sorts servers per human player count, and then per bot count, so an empty server with bots will be lower in server list than a server with one human player (and zero bots).

**Comment by Dune-jr** [comment]
> > If I remember correctly, I was the one who asked for bots to be shown by default. The logic behind this was that players should discover that there are servers with bots that they can play on, before they decide they are not interested in them and hide them. So in the end I'm not sure what the best course of action is here. Opinions?
> 
> In complement with a separate bot count, we can tweak the sort such that it sorts servers per human player count, and then per bot count, so an empty server with bots will be lower in server list than a server with one human player (and zero bots).

Sounds great. I think we should go with this. Might even want to gray out servers that have no non-NPC servers (like empty servers)

**Comment by Dune-jr** [comment]
Additionally, when I suggest to extend this to spectators and sort the server list

- by active players first
- then by spectators
- then by bots

and to gray out the player count of servers with no active players (currently, only grayed out if 0 player). A server that only has spectators is not an interesting server to join. They would still be shown above actually empty servers.
